### PR TITLE
Simplify predicate and script

### LINKED ISSUE #18: Update message to contract predicate with new message retry spec
The latest fuel spec now allows for messages to be natively retryable from the protocol level. This means the predicate no longer needs to check for the correct inputs/outputs or a gas minimum. It only needs to check the inclusion of a specific transaction script (one that calls "process_message" on the target contract)

### PR DESCRIPTION
**Pull request by pixelcircuits**

### CODE REVIEW

File: contract-message-predicate/Forc.toml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 [project]
 authors = ["Fuel Labs <contact@fuel.sh>"]
-entry = "contract_message_predicate.sw"
+entry = "contract_message_test.sw"
```
Comment (Braqzen) [inline contract-message-predicate/Forc.toml:3]: Predicate renamed to test?

File: contract-message-predicate/tests/harness.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            tx.inputs().clone(),
+            tx.outputs().clone(),
+            vec![],
+        );
+
+        // Note: tx inputs[message, contract, coin], tx outputs[contract, change, variable]
+        let _receipts = env::sign_and_call_tx(&wallet, &mut modified_tx).await;
+    }
+    /*
+        #[tokio::test]
```
Comment (Braqzen) [inline contract-message-predicate/tests/harness.rs]: Commented out test?

**Review by Braqzen — COMMENTED** [review]
I don't really know what I'm looking at when it comes to predicates.

File: contract-message-predicate/tests/harness.rs (hunk 2 above)
Comment (pixelcircuits) [reply contract-message-predicate/tests/harness.rs]: WIP

File: contract-message-predicate/Forc.toml (hunk 1 above)
Comment (pixelcircuits) [reply contract-message-predicate/Forc.toml:3]: The only sway files are for tests now

**Comment by pixelcircuits** [comment]
> I don't really know what I'm looking at when it comes to predicates.

I'll definitely be scheduling a meeting to go over this. A LOT is changed in this repo now

File: README.md [hunk 3]
```diff
@@ -1,30 +1,16 @@
...
-
-If all of these conditions are met, then the predicate evaluates as true.
+The Message to Contract Predicate is for messages that are trying to send a data payload to a designated Fuel contract. This predicate simply verifies that the script bytecode hash for the transaction matches for the designated [Message to Contract Script](#message-to-contract-script). If this condition is met, then the predicate evaluates as true.
 
 ### Message to Contract Script
 
 The message to contract predicate relies on a script that performs only the following operations:
 - Transfer the exact amount of base asset specified in the `InputMessage` `amount` field to the contract with ID that matches the first 32 bytes in the message data field
-- Call the function `processMessage` on the contract with ID that matches the first 32 bytes in the message data field
+- Call the function `process_message` on the contract with ID that matches the first 32 bytes in the message data field
```
Comment (nfurfaro) [inline README.md]: As these 2 steps both happen during the call to `process_message`, it might be better to reword this to something like: 
```
Call the function `process_message` on the contract with ID that matches the first 32 bytes in the message data field, and forward the exact amount of base asset specified in the `InputMessage` `amount` field.
```

File: contract-message-predicate/src/contract_message_test.sw [hunk 4]
```diff
@@ -16,6 +12,31 @@ storage {
...
+    #[storage(read)]
+    fn get_test_data2() -> u64;
+    #[storage(read)]
+    fn get_test_data3() -> b256;
+    #[storage(read)]
+    fn get_test_data4() -> Address;
+}
+
+// Get the length of a message input data
+pub fn input_message_data_length(index: u64) -> u64 {
```
Comment (nfurfaro) [inline contract-message-predicate/src/contract_message_test.sw]: This function is now in the stdlib.

File: contract-message-predicate/src/contract_message_test.sw [hunk 5]
```diff
@@ -16,6 +12,31 @@ storage {
...
+    __gtf::<u64>(index, GTF_INPUT_MESSAGE_DATA_LENGTH)
+}
+
+// Get the data of a message input
+pub fn input_message_data<T>(index: u64, offset: u64) -> T {
```
Comment (nfurfaro) [inline contract-message-predicate/src/contract_message_test.sw]: Same with this one, though it now returns a `Bytes` type. I refactored the Bridge-fungible-token contract to handle the new return type already [here](https://github.com/FuelLabs/bridge-fungible-token/blob/f214595fead2b40820a117a49248ae4465ff11d3/bridge-fungible-token/src/utils.sw#L144).

File: contract-message-predicate/src/predicate_asm.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+const GTF_SCRIPT_INPUTS_COUNT: u16 = 0x007;
+const GTF_INPUT_TYPE: u16 = 0x101;
+const GTF_MSG_DATA_LEN: u16 = 0x11A;
+
+const INPUT_MESSAGE_TYPE: u16 = 2;
+const INSTR_PER_WORD: u16 = 2;
+const BYTES_PER_INSTR: u16 = 4;
+
+// Gets the bytecode for the message-to-contract predicate
+pub fn bytecode() -> Vec<u8> {
```
Comment (nfurfaro) [inline contract-message-predicate/src/predicate_asm.rs:13]: This is doable with the Rust SDK out of the box if it would simplify things here a bit.

**Review by nfurfaro — COMMENTED** [review]
Just a few comments, looking good overall though.

File: contract-message-predicate/src/contract_message_test.sw (hunk 5 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/contract_message_test.sw]: How can I convert the Bytes to a u64?

Comment (nfurfaro) [reply contract-message-predicate/src/contract_message_test.sw]: Hmm, we'd have to build a local conversion function based on the into() method for Bytes that converts to b256.
Ideally we would just impl TryInto <u64>() for Bytes, but we don't have that trait yet (blocked).

Comment (nfurfaro) [reply contract-message-predicate/src/contract_message_test.sw]: @pixelcircuits this is what I was refering to here:
```
// NOTE: this cas be lossy! Added here as the From trait currently requires it,
    // but the conversion from `Bytes` ->`b256` should be implemented as
    // `impl TryFrom<Bytes> for b256` when the `TryFrom` trait lands:
    // https://github.com/FuelLabs/sway/pull/3881
    fn into(self) -> b256 {
        let mut value = 0x0000000000000000000000000000000000000000000000000000000000000000;
        let ptr = __addr_of(value);
        self.buf.ptr().copy_to::<b256>(ptr, 1);

        value
    }
```

I think implementing something similar as a free function, i.e: `fn into_u64(b: Bytes) -> u64;` is the way to go for now.

Comment (pixelcircuits) [reply contract-message-predicate/src/contract_message_test.sw]: Bytes to b256 is already supported for `into` [see here](https://github.com/FuelLabs/sway/blob/9934103647eec6eef2db5b56cf51f62a8984d796/sway-lib-std/src/b256.sw). Do you mean an `into` for Bytes to u64? I added a helper function to do just that, but were should it eventually get included in the sway std lib? The conversion to b256 is in the b256.sw file

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -75,18 +65,18 @@ jobs:
...
-        run: forc build --path ${{ matrix.project }}
+        run: forc build
 
       - name: Check Rust formatting
-        run: sh ${{ env.PATH_TO_SCRIPTS }}/fmt.sh ${{ matrix.project }}
+        run: cargo fmt --verbose --check
           
       - name: Run tests
-        run: sh ${{ env.PATH_TO_SCRIPTS }}/test.sh ${{ matrix.project }}
+        run: sh ${{ env.PATH_TO_SCRIPTS }}/test.sh
```
Comment (Braqzen) [inline .github/workflows/ci.yml:79]: Line 76 above builds the contracts. The `test.sh` has 2 commands: build sway then cargo test. No need to build twice, can run cargo test directly here.

File: contract-message-predicate/src/contract_message_test.sw [hunk 8]
```diff
@@ -16,6 +13,29 @@ storage {
     data4: Address = Address::from(ZERO_B256),
 }
 
+// Define verification abi
+abi VerifyMessageData {
+    #[storage(read)]
+    fn get_test_counter() -> u64;
```
Comment (Braqzen) [inline contract-message-predicate/src/contract_message_test.sw]: Not the end of the world but in Rust, and thus in Sway, the convention is to omit the `get_` prefix

File: contract-message-predicate/src/contract_message_test.sw [hunk 9]
```diff
@@ -25,20 +45,20 @@ impl MessageReceiver for Contract {
 
         // Parse the message data
         let data_length = input_message_data_length(msg_idx);
-        if (data_length >= 32) {
-            let contract_id: b256 = input_message_data(msg_idx, 0);
+        if (data_length >= 32u16) {
```
Comment (Braqzen) [inline contract-message-predicate/src/contract_message_test.sw:48]: It's probably better to leave it in; however, in Sway data is packed to be 1 word if it is smaller than 1 word so designating a smaller type doesn't do anything afaik. hm, makes me question why they even exist if that is the case.

File: contract-message-predicate/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+// Make the script and predicate bytecode public
+pub use predicate_asm::bytecode as predicate_bytecode;
+pub use script_asm::bytecode as script_bytecode;
+
+// Gets the hash of the message-to-contract script
+pub fn script_hash() -> [u8; 32] {
+    let script = script_asm::bytecode();
+    let mut script_hasher = Sha256::new();
+    script_hasher.update(script.clone());
```
Comment (Braqzen) [inline contract-message-predicate/src/lib.rs]:
```suggestion
    script_hasher.update(script);
```

File: contract-message-predicate/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    //output to console and build files
+    let script_hash = contract_message_predicate::script_hash();
+    let predicate_root = contract_message_predicate::predicate_root();
+    println!("Script bytecode size is {} bytes.", script.len());
+    println!("Script hash: 0x{}", Bytes32::from(script_hash));
+    println!("Predicate bytecode size is {} bytes.", predicate.len());
+    println!("Predicate root: 0x{}", Address::from(predicate_root));
+    fs::create_dir_all(OUTPUT_DIR).unwrap();
+    fs::write(Path::new(SCRIPT_BUILD_PATH), script).unwrap();
+    fs::write(Path::new(PREDICATE_BUILD_PATH), predicate).unwrap();
```
Comment (Braqzen) [inline contract-message-predicate/src/main.rs]: You may want to consider some error handling if these fail.
Unwrap will panic if there is an issue rather than failing gracefully.
One alternative is to use `.expect("<your message for failure>")`.
Another is to check the result and handle the error if e.g. you cannot create the dir because it already exists (assuming that `create_dir_all` just panics if it exists rather than not doing anything).

File: contract-message-predicate/src/predicate_asm.rs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        op::ret(ONE),
+        //PREDICATE_FAILURE:
+        op::lw(ERR_CODE, INSTR_START, 28 / INSTR_PER_WORD), //ERR_CODE = last word of reference at the end of the program
+        op::rvrt(ERR_CODE),
+        //referenced data (expected script hash, error code return)
+        //00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
+        //FFFFFFFF FFFF0004
+    ]
+    .into_iter()
+    .collect();
```
Comment (Braqzen) [inline contract-message-predicate/src/predicate_asm.rs:75]: Are the opcodes not `u8`s? If they are then `into_iter().collect()` may not be needed.

File: contract-message-predicate/tests/utils/builder.rs [hunk 13]
```diff
@@ -0,0 +1,81 @@
+use std::collections::HashMap;
+
+use fuel_core_interfaces::common::prelude::Word;
+use fuels::prelude::*;
```
Comment (Braqzen) [inline contract-message-predicate/tests/utils/builder.rs:4]: I would import explicitly i.e. comment out, build, see what it complains about and then add it back in manually.

File: contract-message-predicate/tests/utils/builder.rs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    inputs: &[Input],
+    outputs: &[Output],
+    params: TxParameters,
+) -> Script {
+    // Get the script and predicate for contract messages
+    let script_bytecode = contract_message_predicate::script_bytecode();
+
+    // Start building list of inputs and outputs
+    let mut tx_outputs: Vec<Output> = Vec::new();
+    let mut tx_inputs: Vec<Input> = Vec::new();
```
Comment (Braqzen) [inline contract-message-predicate/tests/utils/builder.rs]: It looks like all inputs are added into the vec so instead of creating an empty vec you can create one with the capacity being equal to the length of the number of inputs plus 1 to account for the message on line 25 below. This way you can add into the vector without reallocating multiple times.

File: contract-message-predicate/src/contract_message_test.sw (hunk 9 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/contract_message_test.sw:48]: I assumed they exist for compile time reasons. It could give you a heads up at compile time of possible issues with the developers expectations

File: contract-message-predicate/src/main.rs (hunk 11 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/main.rs]: `create_dir_all` doesn't panic if the dir already exists, but I agree that `.expect` is the better way to go.

File: contract-message-predicate/src/predicate_asm.rs (hunk 12 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/predicate_asm.rs:75]: the opcodes are `Instructions`, which are each 4 bytes long.

File: contract-message-predicate/tests/utils/builder.rs (hunk 13 above)
Comment (pixelcircuits) [reply contract-message-predicate/tests/utils/builder.rs:4]: I tend to agree, but doesn't that defeat the whole reason we have the prelude?

**Review by Braqzen — DISMISSED** [review]
Uncertain about the new codes but looks fine otherwise

File: contract-message-predicate/src/predicate_asm.rs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    const SCRIPT_HASH_PTR: u8 = 0x10;
+    const SCRIPT_PTR: u8 = 0x11;
+    const SCRIPT_LEN: u8 = 0x12;
+    const EXPECTED_HASH_PTR: u8 = 0x13;
+    const COMPARE_RESULT: u8 = 0x14;
+    const VAL_32: u8 = 0x16;
+    const INPUT_INDEX: u8 = 0x17;
+    const INPUT_TYPE: u8 = 0x18;
+    const INPUT_MSG_DATA_LEN: u8 = 0x19;
+    const EXPECTED_INPUT_TYPE: u8 = 0x20;
```
Comment (otrho) [inline contract-message-predicate/src/predicate_asm.rs]: I would argue that `RegId::ONE` is more readable than just `ONE`.  And/or that each of the registers (`0x10` to `0x20`) have a `_REG` suffix, to disambiguate them from the other consts above, like `BYTES_PER_INSTR`.

It doesn't matter, but the next register after `0x19` is `0x1a`. :wink:

File: contract-message-predicate/src/predicate_asm.rs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        op::jnei(INPUT_MSG_DATA_LEN, ZERO, 18), //jumps to PREDICATE_FAILURE if INPUT_MSG_DATA_LEN does not equal 0
+        //SKIP_DATA_CHECK:
+        op::jnei(INPUT_INDEX, ONE, 11), //jumps back to LOOP_START if INPUT_INDEX does not equal 1
+        op::ret(ONE),
+        //PREDICATE_FAILURE:
+        op::ret(ZERO),
+        //referenced data (expected script hash)
+        //00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
+    ]
+    .into_iter()
```
Comment (otrho) [inline contract-message-predicate/src/predicate_asm.rs:74]: I'll put some comments here below so not to pollute the script itself.
- In general it'd be great to have those jump destination 'addresses' (18, 16, 11) as constants as well, with names which match the comment before them.  And if possible, to do a check after collecting them that each offset points the the instruction you expect it to.  That might be overkill for a script which doesn't change often, but a thought.  The 19 could be verified very easily too.
- The `addi(EXPECTED_INPUT_TYPE, ZERO, INPUT_MESSAGE_TYPE)` could just be a `movi`.  You might've copied that from the compiler output.  We had `ADDI` before `MOVI` and some parts of the compiler haven't been updated.
- Similarly the `op::jnei(INPUT_MSG_DATA_LEN, ZERO, 18)` could be `jnzi`.
- You're assuming that `INPUT_INDEX` starts at at least 2, although I guess if it's 1 then it'll be checking the first input which will never have zero length?  And if it's 0 then the `subi` will underflow.
- I've been trying to think of a way to shorten it by optimising the control flow but I don't think there's a way to remove any more instructions. :+1:  Although I did just notice that you could use the heap instead of the stack for the script hash buffer, which, once the new `ALOC` semantics are merged, would mean you could just use `$hp` instead of `SCRIPT_HASH_PTR` and that would remove _one_ `move` instruction. :slightly_smiling_face:

File: contract-message-predicate/src/script_asm.rs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        op::mcpi(CALL_DATA_PTR, CONTRACT_ADDR_PTR, 32), //32 bytes at CALL_DATA_PTR = the 32 bytes at CONTRACT_ADDR_PTR
+        op::mcpi(CALL_DATA_FN_SEL_PTR, FN_SELECTOR_PTR, 4), //4 bytes at CALL_DATA_FN_SEL_PTR = the 4 bytes at FN_SELECTOR_PTR
+        //make contract call
+        op::call(CALL_DATA_PTR, MSG_AMOUNT, ASSET_ID_PTR, CGAS),
+        op::ret(ZERO),
+        //referenced data (function selector)
+        //00000000
+    ]
+    .into_iter()
+    .collect();
```
Comment (otrho) [inline contract-message-predicate/src/script_asm.rs:55]: I think the fact that you're assuming the stack allocated memory to be zero, and so parts of the `CALL_DATA_PTR` are set to zero should be mentioned somewhere, especially about half the selector and the user arg.

And I assume it's tested and working -- I think the selector is treated as a number and so it needs to be big-endian, and so copying it around as bytes might get it backwards.  But if it's working already :+1:

File: contract-message-predicate/src/contract_message_test.sw (hunk 5 above)
Comment (nfurfaro) [reply contract-message-predicate/src/contract_message_test.sw]: Eventually it should be added via implementing TryInto<Bytes> for u64, but this can't be done yet. The workaround is to impl From<u64> for Bytes and use the `into()`method, but this is also blocked atm.

File: contract-message-predicate/src/predicate_asm.rs (hunk 15 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/predicate_asm.rs]: haha. I can't believe I did that, lol

File: contract-message-predicate/src/predicate_asm.rs (hunk 16 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/predicate_asm.rs:74]: Great thinking with using the heap vs the stack!

Comment (pixelcircuits) [reply contract-message-predicate/src/predicate_asm.rs:74]: I spoke too soon... there is no immediate version of `ALOC` like there is `CFEI` so it would require at least 2 instructions to increase the heap.

File: contract-message-predicate/src/script_asm.rs (hunk 17 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/script_asm.rs:55]: The code assumes that memory is initialized to 0 at VM initialization. I think this has to be true otherwise it would be very difficult to sync a node and make proving execution validity on-chain more complex

File: contract-message-predicate/src/predicate_asm.rs (hunk 16 above)
Comment (pixelcircuits) [reply contract-message-predicate/src/predicate_asm.rs:74]: I moved the jump instruction values to constants since some of them are referenced more than once. We could add a check for things like the value for the end of the program code before adding referenced data bits, but we effectively already are indirectly through the tests (particularly the test where everything has to work correctly).

Comment (nfurfaro) [reply contract-message-predicate/src/predicate_asm.rs:74]: >once the new ALOC semantics are merged

Using the heap may be an optimization that has to wait, not sure when it will land.

Using `jnzi` sounds like an easy win though.
I don't have anything to add; @otrho was too thorough :)

Comment (otrho) [reply contract-message-predicate/src/predicate_asm.rs:74]: > I spoke too soon... there is no immediate version of `ALOC` like there is `CFEI` so it would require at least 2 instructions to increase the heap.

Ahh, yeah.  Bugger.  OK, I think it has been golfed all the way.

**Review by otrho — DISMISSED** [review]
Happy with the ASM blocks.

**Review by Voxelot — APPROVED** [review]
